File: lib/buildStyle.ts
import taxonomyJson from "@/data/industry_build_styles.json";

export type BuildStyle = "broad" | "growth";
export type StyleAffinity = "broad" | "growth" | "both" | "unknown";

type StyleTaxonomy = {
  growth: string[];
  broad: string[];
  both: string[];
  growthKeywords?: string[];
  broadKeywords?: string[];
  bothKeywords?: string[];
};

const raw = taxonomyJson as StyleTaxonomy;

const taxonomy = {
  growth: (raw.growth ?? []).map(normIndustry),
  broad: (raw.broad ?? []).map(normIndustry),
  both: (raw.both ?? []).map(normIndustry),
  growthKeywords: raw.growthKeywords ?? [],
  broadKeywords: raw.broadKeywords ?? [],
  bothKeywords: raw.bothKeywords ?? [],
};

function normIndustry(s: string): string {
  return s.trim().toLowerCase();
}

function keywordHit(industry: string, keys: string[] | undefined): boolean {
  if (!keys?.length) return false;
  return keys.some((k) => industry.includes(k.toLowerCase()));
}

/**
 * Classify a Finviz / StockAnalysis industry into build-style affinity.
 * Exact list match first, then keyword fallback for unseen industries.
 */
export function industryStyleAffinity(
  industry?: string | null,
  sector?: string | null
): StyleAffinity {
  const ind = industry ? normIndustry(industry) : "";
  if (ind) {
    if (taxonomy.both.includes(ind)) return "both";
    if (taxonomy.growth.includes(ind)) return "growth";
    if (taxonomy.broad.includes(ind)) return "broad";
    if (keywordHit(ind, taxonomy.bothKeywords)) return "both";
    if (keywordHit(ind, taxonomy.growthKeywords)) return "growth";
    if (keywordHit(ind, taxonomy.broadKeywords)) return "broad";
  }

  // Sector fallback when industry is missing/unknown
  const sec = (sector || "").toLowerCase();
  if (
    sec.includes("information technology") ||
    sec.includes("communication") ||
    sec === "consumer discretionary"
  ) {
    return "growth";
  }
  if (
    sec.includes("utilit") ||
    sec.includes("energy") ||
    sec.includes("real estate") ||
    sec.includes("financial") ||
    sec.includes("consumer staples") ||
    sec.includes("material") ||
    sec.includes("industrial")
  ) {
    return "broad";
  }
  if (sec.includes("health")) return "both";
  return "unknown";
}

/** Whether a stock may fill a Broad or Growth book (user picks always allowed separately). */
export function matchesBuildStyle(
  style: BuildStyle,
  industry?: string | null,
  sector?: string | null
): boolean {
  const aff = industryStyleAffinity(industry, sector);
  if (aff === "both" || aff === "unknown") return true;
  return aff === style;
}

/** Score tilt: higher = better fit for this style. */
export function styleFitBonus(
  style: BuildStyle,
  industry?: string | null,
  sector?: string | null
): number {
  const aff = industryStyleAffinity(industry, sector);
  if (aff === "both") return 0.18;
  if (aff === style) return 0.28;
  if (aff === "unknown") return 0.05;
  return -0.22; // opposite bucket
}

export function parseBuildStyle(raw: unknown): BuildStyle | null {
  const s = String(raw ?? "")
    .trim()
    .toLowerCase();
  if (s === "broad" || s === "growth") return s;
  return null;
}

File: valuation/build_style.py
"""Industry → Broad / Growth / Both affinity (Finviz / StockAnalysis labels).

Reads ``data/industry_build_styles.json`` so TS and Python stay aligned.
"""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Literal, Optional

BuildStyle = Literal["broad", "growth"]
StyleAffinity = Literal["broad", "growth", "both", "unknown"]

_REPO_ROOT = Path(__file__).resolve().parent.parent
_TAXONOMY_PATH = _REPO_ROOT / "data" / "industry_build_styles.json"


@lru_cache(maxsize=1)
def _taxonomy() -> dict:
    raw = json.loads(_TAXONOMY_PATH.read_text(encoding="utf-8"))
    return {
        "growth": {str(x).strip().lower() for x in raw.get("growth", [])},
        "broad": {str(x).strip().lower() for x in raw.get("broad", [])},
        "both": {str(x).strip().lower() for x in raw.get("both", [])},
        "growthKeywords": [str(x).lower() for x in raw.get("growthKeywords", [])],
        "broadKeywords": [str(x).lower() for x in raw.get("broadKeywords", [])],
        "bothKeywords": [str(x).lower() for x in raw.get("bothKeywords", [])],
    }


def _norm(s: Optional[str]) -> str:
    return (s or "").strip().lower()


def _keyword_hit(industry: str, keys: list[str]) -> bool:
    return any(k in industry for k in keys)


def industry_style_affinity(
    industry: Optional[str] = None,
    sector: Optional[str] = None,
) -> StyleAffinity:
    tax = _taxonomy()
    ind = _norm(industry)
    if ind:
        if ind in tax["both"]:
            return "both"
        if ind in tax["growth"]:
            return "growth"
        if ind in tax["broad"]:
            return "broad"
        if _keyword_hit(ind, tax["bothKeywords"]):
            return "both"
        if _keyword_hit(ind, tax["growthKeywords"]):
            return "growth"
        if _keyword_hit(ind, tax["broadKeywords"]):
            return "broad"

    sec = _norm(sector)
    if any(x in sec for x in ("information technology", "communication")) or sec == "consumer discretionary":
        return "growth"
    if any(
        x in sec
        for x in (
            "utilit",
            "energy",
            "real estate",
            "financial",
            "consumer staples",
            "material",
            "industrial",
        )
    ):
        return "broad"
    if "health" in sec:
        return "both"
    return "unknown"


def matches_build_style(
    style: BuildStyle,
    industry: Optional[str] = None,
    sector: Optional[str] = None,
) -> bool:
    aff = industry_style_affinity(industry, sector)
    if aff in ("both", "unknown"):
        return True
    return aff == style


def style_fit_bonus(
    style: BuildStyle,
    industry: Optional[str] = None,
    sector: Optional[str] = None,
) -> float:
    aff = industry_style_affinity(industry, sector)
    if aff == "both":
        return 0.18
    if aff == style:
        return 0.28
    if aff == "unknown":
        return 0.05
    return -0.22


def parse_build_style(raw: Optional[str]) -> Optional[BuildStyle]:
    if not raw:
        return None
    s = str(raw).strip().lower()
    if s in ("broad", "growth"):
        return s  # type: ignore[return-value]
    return None
